internal/pkg/service/buffer/sink/tablesink/storage/repository/file_repo.go

```diff
@@ -153,16 +205,212 @@ func (r *FileRepository) Delete(k storage.FileKey) *op.TxnOp {
 	return txn
 }
 
-// create saves a new entity, see also update method.
+// rotate one file, it is a special case of the rotateAllIn.
+func (r *FileRepository) rotate(rb rollback.Builder, now time.Time, sinkKey key.SinkKey, sink *definition.Sink, openNewFile bool) *op.AtomicOp[storage.File] {
+	var file storage.File
+	var sinks []definition.Sink
+	return op.Atomic(r.client, &file).
+		BeforeWrite(func(context.Context) {
+			// Create sinks slice before write, after the read phase, because the sink value may not be available sooner
+			sinks = []definition.Sink{*sink}
+		}).
+		AddFrom(r.
+			rotateAllIn(rb, now, sinkKey, &sinks, openNewFile).
+			AddProcessor(func(_ context.Context, result *op.Result[[]storage.File]) {
+				// Unwrap results, there in only one file
+				if result.Err() == nil {
+					files := result.Result()
+					if count := len(files); count == 1 {
+						file = files[0]
+					} else {
+						result.AddErr(errors.Errorf(`expected 1 file, found %d`, count))
+					}
+				}
+			}),
+		)
+}
+
+// rotateAllIn is a common code used by rotate and close operations.
+func (r *FileRepository) rotateAllIn(rb rollback.Builder, now time.Time, parentKey fmt.Stringer, sinksPtr *[]definition.Sink, openNew bool) *op.AtomicOp[[]storage.File] {
```

Comment:
If you have an idea how to do it better, I will be glad.
I would like to call about this.

It is necessary to realize that this is the key and complex logic of the entire system, written in my opinion quite readable and short.

By dividing it into smaller parts, you lose an overview of what is happening there.

It should also be considered that it must be one etcd transaction to make it atomic.

I will send a dump of the entire etcd transaction here, it may be interesting for discussion.